src/token-providing-subscription.ts

```diff
@@ -1,147 +1,161 @@
-                                            }
-                                        },
-                                        onEvent: listeners.onEvent,
-                                        onEnd: error => {
-                                            onTransition(new EndedSubscriptionState(error));
-                                        }
-                                    },
-                                    headers
-                                )
-                            })
-                            .catch( error => {
-                                error => {
-                                    onTransition(new FailedSubscriptionState(error));
-                                }
-                            });
-                    }
-                    fetchTokenAndExecuteSubscription();
-                }
-                
-                unsubscribe(){
-                    if(this.tokenPromise) this.tokenPromise.cancel();
-                    this.underlyingSubscription.unsubscribe();
-                    this.onTransition(new EndedSubscriptionState());
-                }
-
-                private putTokenIntoHeader(token: any) {
-                    if(token) {
-                        headers['Authorization'] = `Bearer ${token}`;
-                        logger.verbose(`TokenProvidingSubscription: token fetched: ${token}`);
-                    }
-                }               
-            }
+export let createTokenProvidingStrategy: (
+  nextSubscribeStrategy: SubscribeStrategy,
+  logger: Logger,
+  tokenProvider?: TokenProvider
+) => SubscribeStrategy = (nextSubscribeStrategy, logger, tokenProvider) => {
+  // Token provider might not be provided. If missing, go straight to underlying subscribe strategy.
+  if (tokenProvider) {
+    return (listeners, headers) =>
+      new TokenProvidingSubscription(
+        logger,
+        listeners,
+        headers,
+        tokenProvider,
+        nextSubscribeStrategy
+      );
+  } else {
+    return (listeners, headers) =>
+      nextSubscribeStrategy(listeners, headers);
+  }
+}
 
-            class OpenSubscriptionState implements SubscriptionState {
-                constructor(private headers: ElementsHeaders, private underlyingSubscription: Subscription, private onTransition: (SubscriptionState) => void){
-                    logger.verbose(`TokenProvidingSubscription: transitioning to OpenSubscriptionState`);
-                    listeners.onOpen(headers);
-                }
+export interface TokenProvidingSubscriptionState {
+  subscribe(token: string, listeners: SubscriptionListeners): void;
+  unsubscribe(): void;
+}
 
-                unsubscribe(){
-                    this.underlyingSubscription.unsubscribe();
-                    this.onTransition(new EndedSubscriptionState());
-                }
+class TokenProvidingSubscription implements Subscription {
+  private state: TokenProvidingSubscriptionState;
+  private tokenPromise: Promise<any>;
```

Comment:
You don't need `tokenPromise` to be a property. It's only used in `subscribe` method scope.